src/schedule/timer_domain.c

```diff
@@ -231,9 +231,20 @@ static void timer_domain_set(struct ll_schedule_domain *domain, uint64_t start)
 	uint64_t ticks_tout = domain->ticks_per_ms * timer_domain->timeout /
 			     1000;
 	uint64_t ticks_set;
+	uint64_t ticks_now = platform_timer_get(timer_domain->timer);
 #ifndef __ZEPHYR__
 	uint64_t ticks_req = start + ticks_tout;
 
+	/* check we are not scheduling in the past ? */
+	if (ticks_req <= ticks_now) {
+		/* LL work has overshot the LL period duration meaning it's
+		 * taking too long to complete the work. Reschedule in 1 periods
+		 * time and let driver/userspace deal with the delay.
+		 */
+		tr_err(&ll_tr, "timer_domain_set(): overshoot by %d ticks",
+		       (uint32_t)(ticks_now - ticks_req));
+		ticks_req = ticks_now + ticks_tout;
```

Comment:
try to launch it ASAP, not after 1ms? e.g. 'ticks_req = ticks_now + 200;', actually we already has similar check:
```
int64_t platform_timer_set(struct timer *timer, uint64_t ticks)
{
	/* a tick value of 0 will not generate an IRQ */
	if (ticks == 0)
		ticks = 1;

	/* Check if requested time is not past time */
	if (ticks > shim_read64(SHIM_DSPWC))
		shim_write64(SHIM_DSPWCT0C, ticks);
	else
		shim_write64(SHIM_DSPWCT0C, shim_read64(SHIM_DSPWC) +
			     TIMER_MIN_RECOVER_CYCLES); //Liam, we already has check and add 100 if ticks <=now.

	/* Enable IRQ */
	shim_write(SHIM_DSPWCTCS, SHIM_DSPWCTCS_T0A);

	return shim_read64(SHIM_DSPWCT0C);
}
```

Reply:
@keyonjie this is actually not optimal, as the timer driver has no client context to make any adjustments. timer should really return an error here and allow client caller to specify a new time.  Btw, it also looks like the int64_t return value should be a uint64_t since we are returning a 64bit counter ?? i.e. I would rework the API as
```c
int platform_timer_set(struct timer *timer, uint64_t ticks, uint64_t *ticks_set);
```
Where tick_set can be used by clients for getting the time actually set and return an error with negative return value

Reply:
Going to remove the adjustment here, so that we only get the error messages.